### PR TITLE
MediaScrubProgressBar: Replacement for TimeNavigationTitleView position slider

### PR DESCRIPTION
**Pull request by robwayne**
### Checklist
- [x] I've run `bundle exec fastlane test` from the root directory to see all new and existing tests pass
- [x] I've followed the [vlc-ios code style](Docs/CodingStyle.md)
- [x] I've read the [Contribution Guidelines](https://github.com/videolan/vlc-ios#contribute)
- [x] I've updated the documentation if necessary.

### CODE REVIEW

File: Sources/MediaScrubProgressBar.swift [hunk 1]
```diff
@@ -59,33 +68,45 @@ protocol MediaScrubProgressBarDelegate {
...
+        spacing = 5
+        axis = .vertical
+        translatesAutoresizingMaskIntoConstraints = false
+    }
+
+    @objc func setRemainingTimeLabelText(_ text: String) {
+        remainingTimeLabel.text = text
+    }
+
+    @objc func setElapsedTimeLabelText(_ text: String) {
```
Comment (carolanitz) [inline Sources/MediaScrubProgressBar.swift]: not needed have an update method that gets the information from vpc directly

File: Sources/VLCMovieViewController.m [hunk 2]
```diff
@@ -371,10 +373,30 @@ - (void)setupConstraints
...
         constraints = [constraints arrayByAddingObjectsFromArray:[self getVideoOptionsConstraints]];
+        constraints = [constraints arrayByAddingObjectsFromArray:[self getScrubProgressBarConstraints]];
     #endif
     [NSLayoutConstraint activateConstraints: constraints];
 }
 
+- (NSArray *)getScrubProgressBarConstraints
+{
+    float margin = 20.0f;
+    UILayoutGuide *guide;
```
Comment (carolanitz) [inline Sources/VLCMovieViewController.m]: set the guide directly to self.view.layoutmarginsguide

File: Sources/VLCMovieViewController.m [hunk 3]
```diff
@@ -1121,11 +1150,9 @@ - (void)volumeSliderAction:(id)sender
 
 - (void)updateTimeDisplayButton
 {
+    NSString *elapsedTime = _displayRemainingTime ? [[_vpc remainingTime] stringValue] : [[_vpc playedTime] stringValue];
     UIButton *timeDisplayButton = self.timeNavigationTitleView.timeDisplayButton;
-    if (_displayRemainingTime)
-        [timeDisplayButton setTitle:[[_vpc remainingTime] stringValue] forState:UIControlStateNormal];
-    else
-        [timeDisplayButton setTitle:[[_vpc playedTime] stringValue] forState:UIControlStateNormal];
+    [timeDisplayButton setTitle:elapsedTime forState:UIControlStateNormal];
```
Comment (carolanitz) [inline Sources/VLCMovieViewController.m]: I would put that in a separate commit since it has nothing to do with this PR/ commit message

File: Sources/VLCMovieViewController.m [hunk 4]
```diff
@@ -1891,4 +1925,56 @@ - (void) mediaMoreOptionsActionSheetDidToggleInterfaceLockWithState:(BOOL)state
...
+    [_scrubProgressBar setRemainingTimeLabelText:[[_vpc remainingTime] stringValue]];
+    [_scrubProgressBar setElapsedTimeLabelText:[[_vpc playedTime] stringValue]];
+}
+
+- (void)updatePlaybackPosition
+{
+    if(!_positionSet) {
+        float position = _scrubProgressBar.progressSlider.value;
+        [_vpc setPlaybackPosition:position];
+        [_vpc setNeedsMetadataUpdate];
```
Comment (carolanitz) [inline Sources/VLCMovieViewController.m]: can we move the setNeedsMetadataupdate into the setPlaybackPosition and remove it actually from the header of VLCPlaybackController? I don't think this should be done from the movieviewcontroller in general

File: Sources/MediaScrubProgressBar.swift [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        label.font = UIFont(name: "SFProDisplay-Bold", size: 12)
+        label.textColor = .orange
+        label.text = "000:00"
+        label.translatesAutoresizingMaskIntoConstraints = false
+        return label
+    }()
+    
+    var remainingTimeLabel: UILabel = {
+        var label = UILabel()
+        label.font = UIFont(name: "SFProDisplay-Bold", size: 12)
```
Comment (carolanitz) [inline Sources/MediaScrubProgressBar.swift]: don't use this font directly !! Systemfont is the way to go. if apple changes the standard font this breaks the UI ^

File: Sources/MediaScrubProgressBar.swift [hunk 6]
```diff
@@ -9,45 +9,54 @@
...
-        slider.thumbTintColor = .orange
         slider.isContinuous = true
-        slider.addTarget(self, action: #selector(moveSliderThumb), for: .touchUpInside)
-        slider.translatesAutoresizingMaskIntoConstraints = false
+        slider.addTarget(self, action: #selector(moveSliderThumb), for: .valueChanged)
+        slider.addTarget(self, action: #selector(progressSliderTouchDown), for: .touchDown)
+        slider.addTarget(self, action: #selector(progressSliderTouchUp), for: .touchUpInside)
+        slider.addTarget(self, action: #selector(progressSliderTouchUp), for: .touchUpOutside)
+        slider.addTarget(self, action: #selector(progressSliderDragged), for: .touchDragInside)
+        slider.addTarget(self, action: #selector(progressSliderDragged), for: .touchDragOutside)
```
[the remark this answers is not in the document]
Comment (robwayne) [reply Sources/MediaScrubProgressBar.swift]: The I did, however the valueChanged doesnt catch the dragging outside the bounds of the uislider, therefore it doesnt update the scrubbing information, also the touchup and touchdown events are needed to determine when scrubbing has started versus ended individually

File: Sources/VLCMovieViewController.m [hunk 7]
```diff
@@ -1891,4 +1924,16 @@ - (void) mediaMoreOptionsActionSheetDidToggleInterfaceLockWithState:(BOOL)state
...
+
+- (void)mediaSrubProgressBarShouldResetIdleTimer
+{
+    [self _resetIdleTimer];
+}
+
+- (void)mediaProgressScrubBarDidUpdateScrubbingWithIsScrubbing:(BOOL)isScrubbing
+{
+    self.scrubIndicatorView.hidden = !isScrubbing;
+}
```
Comment (carolanitz) [inline Sources/VLCMovieViewController.m]: all of this should go in the new extension that we talked about ^

File: Sources/MediaScrubProgressBar.swift [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    private func updateScrubLabel() {
+        switch progressSlider.scrubbingSpeed {
+        case 1.0:
+            scrubbingIndicatorLabel.text = NSLocalizedString("PLAYBACK_SCRUB_HIGH", comment:"")
+        case 0.5:
+            scrubbingIndicatorLabel.text = NSLocalizedString("PLAYBACK_SCRUB_HALF", comment: "")
+        case 0.25:
+            scrubbingIndicatorLabel.text = NSLocalizedString("PLAYBACK_SCRUB_QUARTER", comment: "")
+        default:
+            scrubbingIndicatorLabel.text = NSLocalizedString("PLAYBACK_SCRUB_FINE", comment: "")
```
Comment (carolanitz) [inline Sources/MediaScrubProgressBar.swift]: no default please define the exact value range here so that we get a warning when something changes in the range for scrubbingspeed

File: Sources/MediaScrubProgressBar.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (carolanitz) [reply Sources/MediaScrubProgressBar.swift]: hmm okay my gut tells me that this is should be less complex but I can see that this is how we've done it in the xib so it's fine

File: Sources/MediaScrubProgressBar.swift [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+
+    @objc private func updatePlaybackPosition() {
+        if !positionSet {
+            playbackController.playbackPosition = progressSlider.value
+            playbackController.setNeedsMetadataUpdate()
+            positionSet = true
+        }
+    }
+
+    @objc func updateForScrubbing() {
```
Comment (carolanitz) [inline Sources/MediaScrubProgressBar.swift]: can we just name it updateUI because updateforscrubbing sounds like you update in preparation for scrubbing.

File: Sources/MediaScrubProgressBar.swift [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    }
+    
+    // MARK: Slider Methods
+    @objc private func moveSliderThumb() {
+        perform(#selector(updatePlaybackPosition), with: nil, afterDelay: 0.3)
+        if playbackController.mediaDuration > 0 {
+            updateForScrubbing()
+        }
+        positionSet = false
+        delegate?.mediaSrubProgressBarShouldResetIdleTimer()
```
Comment (carolanitz) [inline Sources/MediaScrubProgressBar.swift]: srub

File: Sources/MediaScrubProgressBar.swift [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    @objc private func progressSliderTouchDown() {
+        updateScrubLabel()
+        isScrubbing = true
+        scrubLabelStackView.isHidden = !isScrubbing
+        delegate?.mediaProgressScrubBarDidUpdateScrubbing(isScrubbing: isScrubbing)
```
Comment (carolanitz) [inline Sources/MediaScrubProgressBar.swift]: can we have a did start scrubbing here?

File: Sources/MediaScrubProgressBar.swift [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    @objc private func progressSliderTouchUp() {
+        isScrubbing = false
+        scrubLabelStackView.isHidden = !isScrubbing
+        delegate?.mediaProgressScrubBarDidUpdateScrubbing(isScrubbing: isScrubbing)
```
Comment (carolanitz) [inline Sources/MediaScrubProgressBar.swift]: and a did stop scrubbing here?

File: Sources/MediaScrubProgressBar.swift [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+        slider.maximumValue = 1
+        slider.minimumTrackTintColor = .orange
+        slider.maximumTrackTintColor = UIColor(red: 1, green: 1, blue: 1, alpha: 0.2)
+        slider.setThumbImage(UIImage(named: "sliderThumb"), for: .normal)
+        slider.isContinuous = true
+        slider.addTarget(self, action: #selector(moveSliderThumb), for: .valueChanged)
+        slider.addTarget(self, action: #selector(progressSliderTouchDown), for: .touchDown)
+        slider.addTarget(self, action: #selector(progressSliderTouchUp), for: .touchUpInside)
+        slider.addTarget(self, action: #selector(progressSliderTouchUp), for: .touchUpOutside)
+        slider.addTarget(self, action: #selector(progressSliderDragged), for: .touchDragInside)
```
Comment (carolanitz) [inline Sources/MediaScrubProgressBar.swift]: call updatescrublabel here directly

File: Sources/MediaScrubProgressBar.swift (hunk 11 above)
Comment (robwayne) [reply Sources/MediaScrubProgressBar.swift]: but wouldnt that be creating more biolerplate code, because I'd have two functions that basically just set the scrubActivity indicator to hidden or visible

File: Sources/MediaScrubProgressBar.swift (hunk 12 above)
Comment (robwayne) [reply Sources/MediaScrubProgressBar.swift]: same as above^

**Comment by Mikanbu** [comment]
This PR will be handled by [MR 197](https://code.videolan.org/videolan/vlc-ios/-/merge_requests/197).
